README.md
# Wsus-client Cookbook
[![Cookbook Version][cookbook_version]][supermarket_url]
[![Build Status][build_status]][build_status]
[![License][license]][license]

Configures WSUS clients to retrieve approved updates.

## Testing

The PowerShell script will always fail if run via the winrm vagrant provider
as the IUpdateSession::CreateUpdateDownloader is not available remotely.

## Logs

The Microsoft.Update.Session object keeps a log in: C:\Windows\WindowsUpdate.log

## Requirements

This cookbook recommends Chef 12.6+.

### Platforms
* Windows XP
* Windows Vista
* Windows Server 2003 R2
* Windows 7
* Windows Server 2008 (R1, R2)
* Windows 8 and 8.1
* Windows Server 2012 (R1, R2)
* Windows 10
* Windows Server 2016

## Usage

Using this cookbook is quite easy; add the desired recipes to the run list of a node, or role.
Adjust any attributes as desired. For example, to configure a windows server role that connects to your WSUS server:

```ruby
$ cat roles/updated_windows_server.rb
name 'updated_windows_server'
description 'Setup a windows server to keep up-to-date'

run_list(
  'wsus-client::default'
)

default_attributes(
  wsus_client: {
    wsus_server: 'http://wsus-server.my-corporation.com:8530',
    update_group: 'updated_server2015',
  },
)
```

## Providers & Resources

### update
This provider allows to synchronously download and/or install available windows updates.

#### Actions
Action   | Description
---------|---------------------------
download | Download available updates
install  | Install downloaded updates

> NOTE: The default behavior is `[:download, :install]`

#### Attributes
Attribute        | Description                                            | Type                  | Default
-----------------|--------------------------------------------------------|-----------------------|------------------------
actions          | An array of actions to perform (see. actions above)    | Symbol array          | `[:download, :install]`
name             | Name of the resource                                   | String                |
download_timeout | Time alloted to the download operation before failing  | Integer               | `3600`
install_timeout  | Time alloted to the install operation before failing   | Integer               | `3600`
handle_reboot    | Indicate whether to reboot or not after update install | TrueClass, FalseClass | `false`
reboot_delay     | The amount of time (in minutes) to delay the reboot    | Integer               | `1`

## Recipes

### wsus-client::default
Convenience recipe that configures WSUS client and performs a synchronous update.
It basically includes `wsus-client::configure` and `wsus-client::update`

### wsus-client::configure
This recipe modifies the Windows registry to configure [WSUS update settings][wsus_registry].

#### Attributes
The following attributes are used to configure the `wsus-client::configure` recipe, accessible via `node['wsus_client'][attribute]`.

Attribute                    | Description                                                                                            | Type                | Default
-----------------------------|--------------------------------------------------------------------------------------------------------|---------------------|--------
wsus_server                  |Defines a custom WSUS server to use instead of Microsoft Windows Update server                          |String, URI          |`nil`
update_group                 |Defines the current computer update group. (see [client-side targeting][client_targeting])              |String               |`nil`
disable_windows_update_access|Disables access to Windows Update (or your WSUS server)                                                 |TrueClass, FalseClass|`false`
enable_non_microsoft_updates |Allows signed non-Microsoft updates.                                                                    |TrueClass, FalseClass|`true`
allow_user_to_install_updates|Authorizes Users to approve or disapprove updates.                                                      |TrueClass, FalseClass|`false`
auto_install_minor_updates   |Defines whether minor update should be automatically installed.                                         |TrueClass, FalseClass|`false`
no_reboot_with_logged_users  |Disables automatic reboot with logged-on users.                                                         |TrueClass, FalseClass|`true`
automatic_update_behavior    |Defines auto update behavior.                                                                           |Symbol`*`            |`:disabled`
detection_frequency          |Defines times in hours between detection cycles.                                                        |FixNum               |`22`
schedule_install_day         |Defines the day of the week to schedule update install.                                                 |Symbol`**`           |`:every_day`
schedule_install_time        |Defines the time of day in 24-hour format to schedule update install.                                   |FixNum (1-23)        |`0`
schedule_retry_wait          |Defines the time in minutes to wait at startup before applying update from a missed scheduled time      |FixNum (0-60)        |`0`
reboot_warning               |Defines time in minutes of the restart warning countdown after reboot-required updates automatic install|FixNum (1-30)        |`5`
reboot_prompt_timeout        |Defines time in minutes between prompts for a scheduled restart                                         |FixNum (1-1440)      |`10`
service_timeout              |Defines time in seconds to wait before the service resource times out                                   |Integer              |`60`

`*` automatic_update_behavior values are:

```ruby
# :disabled  = Disables automatic updates
# :detect    = Only notify users of new updates
# :download  = Download updates, but let users install them
# :install   = Download and install updates
# :manual    = Lets the users configure the behavior
```

`**` schedule_install_day possible values are: `:every_day`, `:sunday`, `:monday`, `:tuesday`, `:wednesday`, `:thursday`, `:friday`, `:saturday`

### wsus-client::update

This recipe performs a synchronous detection and install of available Windows updates.

#### Attributes

The following attributes are used to configure the `wsus-client::update` recipe, accessible via `node['wsus_client']['update'][attribute]`.

Attribute        | Description                                             | Type                  | Default
-----------------|---------------------------------------------------------|-----------------------|--------
action           | Define actions performed by the update recipe.          | Array of symbols      | `[:download, :install]`
download_timeout | Time alloted to the download operation before failing.  | Integer               | `3600`
install_timeout  | Time alloted to the install operation before failing.   | Integer               | `3600`
handle_reboot    | Indicate whether to reboot or not after update install. | TrueClass, FalseClass | `false`
reboot_delay     | The amount of time (in minutes) to delay the reboot.    | Integer               | `1`

# Contributing

1. Fork the [repository on Github][repository]
2. Create a named feature branch (like `add_component_x`)
3. Write your change
4. Write tests for your change (if applicable)
5. Run the tests, ensuring they all pass
6. Submit a Pull Request using Github

# License and Authors

Authors: [Baptiste Courtois][author] (<b.courtois@criteo.com>)

```text
Copyright 2014-2015, Criteo.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
```

[author]:                   https://github.com/Annih
[repository]:               https://github.com/criteo-cookbooks/wsus-client
[license]:                  https://img.shields.io/github/license/criteo-cookbooks/wsus-client.svg
[client_targeting]:         https://technet.microsoft.com/library/cc720450
[wsus_registry]:            https://technet.microsoft.com/library/dd939844
[build_status]:             https://api.travis-ci.org/criteo-cookbooks/wsus-client.svg?branch=master
[cookbook_version]:         https://img.shields.io/cookbook/v/wsus-client.svg
[supermarket_url]:          https://supermarket.chef.io/cookbooks/wsus-client


CHANGELOG.md
Wsus-client CHANGELOG
==============
This file is used to list changes made in each version of the wsus-client cookbook.

3.0.0 (2022-02-11)
------------------
Drop support for old Chef versions (< 13).

2.0.3 (2019-07-23)
------------------
- [PR 44](https://github.com/criteo-cookbooks/wsus-client/pull/44) - Add timeout attribute for service resource

2.0.2 (2019-01-10)
------------------
- [PR 37](https://github.com/criteo-cookbooks/wsus-client/pull/37) - Fix for always restarting when some config opts are empty

2.0.1 (2019-01-10)
------------------
- [PR 40](https://github.com/criteo-cookbooks/wsus-client/pull/40) - Retries twice wuauserv restart calls

2.0.0 (2017-10-17)
------------------
- [PR 35](https://github.com/criteo-cookbooks/wsus-client/pull/35) - **Breaking:** Changed attributes controlling the `wsus-client::update` recipe
- [PR 34](https://github.com/criteo-cookbooks/wsus-client/pull/34) - Add progress tracking support to `wsus_client_update` actions
- [PR 33](https://github.com/criteo-cookbooks/wsus-client/pull/33) - Rewrite `wsus_client_update` as a Custom resource
- Configure travis to test using ruby `2.3.0`

1.2.1 (2016-05-05)
------------------
- Configure travis to test using ruby 2.1.0
- [PR 21](https://github.com/criteo-cookbooks/wsus-client/pull/21) - Support chef-client running on ruby 64bits (chef >= 12.8.1)

1.2.0 (2015-08-07)
------------------
- [PR 17](https://github.com/criteo-cookbooks/wsus-client/pull/14) - Speed up LWRP by searching updates only once
- [PR 15](https://github.com/criteo-cookbooks/wsus-client/pull/15) - Log more information before downloading and installing updates

1.1.0 (2015-06-18)
------------------
- Expose new wsus_client_update LWRP to synchronously download or install updates
- [PR 9](https://github.com/criteo-cookbooks/wsus-client/pull/9) - Fix the client-side detection
- [PR 11](https://github.com/criteo-cookbooks/wsus-client/pull/11) - Allow string values for automatic_update_behavior and schedule_install_day
- Add chefspec and rspec tests!

1.0.1 (2015-06-03)
------------------
- Fix README.md for chef supermarket

1.0.0 (2015-06-03)
------------------
- Workaround foodcritic false positive
- Rename attribute `disable_windows_update_acces` to `disable_windows_update_access`
- Explicitly set attribute value of `detection_frequency` to 22 hours
- Force WU detection on configuration change
- Complete README.md

0.1.3 (2015-01-29)
------------------
- Fail chef run on guard failure
- Configure basic rake tasks and travis build

0.1.2 (2014-09-15)
------------------
-  Fix update script guard condition

0.1.1 (2014-09-15)
------------------
- Add windows supports in metadata

0.1.0 (2014-08-27)
------------------
- Initial release of wsus-client
- Client recipes do nothing on non-windows platform
- Add client recipes that configures automatic updates and initiates download and install


attributes/default.rb
#
# Author:: Baptiste Courtois (<b.courtois@criteo.com>)
# Cookbook Name:: wsus-client
# Attribute:: default
#
# Copyright:: Copyright (c) 2014 Criteo.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
# WSUS is a windows feature
return unless platform?('windows')

# Defines whether a custom WSUS server should be used instead of Microsoft Windows Update server.
default['wsus_client']['wsus_server']                              = nil

# Defines the current computer update group.
# => Truthy value also enable client-side update group targeting.
default['wsus_client']['update_group']                             = nil

# Disables access to Windows Update (or your WSUS server).
default['wsus_client']['disable_windows_update_access']            = false
# Allows signed non-Microsoft updates.
default['wsus_client']['enable_non_microsoft_updates']             = true
# Authorizes Users to approve or disapprove updates.
default['wsus_client']['allow_user_to_install_updates']            = false
# Defines whether minor update should be automatically installed.
default['wsus_client']['auto_install_minor_updates']               = false
# Disables automatic reboot with logged-on users.
default['wsus_client']['no_reboot_with_logged_users']              = true

# Defines auto update behavior.
# => :disabled  = Disables automatic updates
# => :detect    = Only notify users of new updates
# => :download  = Download updates, but lets users install them
# => :install   = Download and install updates
# => :manual    = Lets the users configure the behavior
default['wsus_client']['automatic_update_behavior']                = :disabled

# Defines times in hours between detection cycles.
# => 0          = disables custom detection frequency (use Microsoft default value of 22 hours)
# => 1-22       = enables custom detection frequency with specified value
default['wsus_client']['detection_frequency']                      = 22

# Defines the day of the week to schedule update install.
# => :every_day = updates will be installed every day
# => :sunday    = updates will be installed every sunday
# => :monday    = updates will be installed every monday
# => :tuesday   = updates will be installed every tuesday
# => :wednesday = updates will be installed every wednesday
# => :thursday  = updates will be installed every thursday
# => :friday    = updates will be installed every friday
# => :saturday  = updates will be installed every saturday
default['wsus_client']['schedule_install_day']                     = :every_day
# Defines the time of day in 24-hour format to schedule update install. (0-23)
default['wsus_client']['schedule_install_time']                    = 0
# Defines the time in minutes to wait at startup before applying update from a missed scheduled time
# => 0          = attempt the missed installation during the next scheduled installation time
# => 1-60       = time in minutes to wait before applying missed
default['wsus_client']['schedule_retry_wait']                      = 0

# Defines the time in minutes of the restart warning countdown after reboot-required updates automatic install
# => 0          = use default value of 5 minutes
# => 1-30       = set restart countdown to specified value
default['wsus_client']['reboot_warning']                           = 5
# Defines time in minutes between prompts for a scheduled restart
# => 0          = use default value of 10 minutes
# => 1-1440     = set prompts period to specified value
default['wsus_client']['reboot_prompt_timeout']                    = 10

# Defines time in seconds to wait before the service resource times out
default['wsus_client']['service_timeout']                          = 60

# Define action performed by the update recipe
# This can be a combination of: nothing, download & install
default['wsus_client']['update']['action'] = %i[download install]
# Time in seconds alloted to the download operation before failing.
default['wsus_client']['update']['download_timeout']               = 3600
# Time in seconds alloted to the install operation before failing.
default['wsus_client']['update']['install_timeout']                = 3600
# Indicate whether to reboot or not after update install.
default['wsus_client']['update']['handle_reboot']                  = false
# The amount of time (in minutes) to delay the reboot.
default['wsus_client']['update']['reboot_delay']                   = 1
